common/scala/src/main/scala/org/apache/openwhisk/common/Logging.scala

```diff
@@ -414,6 +414,28 @@ object LoggingMarkers {
       LogMarkerToken(loadbalancer + controllerInstance.asString, s"memory${actionType}Inflight", counter)(
         MeasurementUnit.none)
 
+  // Counter metrics for completion acks in load balancer
+  sealed trait completionAckType { def asString: String = ??? }
+  case object regularCompletionAck extends completionAckType { override def asString = "regular" }
+  case object forcedCompletionAck extends completionAckType { override def asString = "forced" }
+  case object healthcheckCompletionAck extends completionAckType { override def asString = "healthcheck" }
+  case object regularAfterForcedCompletionAck extends completionAckType { override def asString = "regularAfterForced" }
+  case object forcedAfterRegularCompletionAck extends completionAckType { override def asString = "forcedAfterRegular" }
```

Comment:
This change would be useful to diagnose operational issues.
I have a few comments.

In a small scope, if you don't have any other purpose to use `toString`, I think you can directly override `toString` here.
Then you can just use that like this: `"type" -> completionAckType`.

And I prefer to use the upper case first character for the object and trait according to the scala convention.
https://docs.scala-lang.org/style/naming-conventions.html


In a bigger scope, this is a good way to define subclass(object) but if this trait is only used here I'd rather define them using just `Sting`.

```scala
def LOADBALANCER_COMPLETION_ACK(controllerInstance: ControllerInstanceId, completionAckType: String) =
    if (TransactionId.metricsKamonTags)
      LogMarkerToken(
        loadbalancer,
        "completionAck",
        counter,
        None,
        Map("controller_id" -> controllerInstance.asString, "type" -> completionAckType))(MeasurementUnit.none)
    else
      LogMarkerToken(
        loadbalancer + controllerInstance.asString,
        "completionAck_" + completionAckType,
        counter)(MeasurementUnit.none)
```

If you mind a client uses this with just string, I think you can also define them as a `Constant`.

```scala
val RegularCompletionAck: String = "regular"
.
.
MetricEmitter.emitCounterMetric(
            LoggingMarkers.LOADBALANCER_COMPLETION_ACK(controllerInstance, RegularCompletionAck))
        }
```

Reply:
@style95 for posterity: We use `asString` throughout the codebase for places where we explicitly want a conversion to string to happen. `toString` way too often bites you so we make these conversions as explicit as possible.

Reply:
oh ok, I didn't know the history.

Reply:
Hm... I introduced `case object`s here for type safety. When using `String`s, you could basically feed in any `String` value. Using pre-defined `String` values is just convenience but no guarantee that "somebody" uses a bogus `String` value you do not expect in your metrics system. Why not use Scala's type system?

I intentionally used `case object`s so they should already be singletons. Or do I miss something here? So there should be no need to define convenience `val`s.

Also see https://github.com/apache/incubator-openwhisk/blob/bf63bb2c31e5907689655d226511cbb99126b734/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerProxy.scala#L43-L52.

@markusthoemmes thanks for jumping in on `asString`. That's exactly why I used `asString` instead of `toString`. If you agree, I would like to keep `asString`.

I agree on the Scala naming convention. Too much golang...

Reply:
> I intentionally used case objects so they should already be singletons. Or do I miss something here? So there should be no need to define convenience vals.

`LogMarkerToken` are meant to be treated like static/singleton as they in turn hold reference to actual metric instance like counter. In current approach we create a new instance of token for every call to `emitCounterMetric`. Further that token instance would then inturn lookup the backing `Kamon` metric instance. 

So to avoid this overhead in metric computation it would be better to refer to tokens as vals. This ensures that `MetricEmitter.emitCounterMetric` call just leads to only increment operation and does not create new instance and do lookup.

Reply:
I had a private sync with @chetanmeh on his proposal.

* Currently, this PR defines a counter metric convenience function `LOADBALANCER_COMPLETION_ACK` in object `LoggingMarkers` which is called over and over again by `CommonLoadBalancer`. This will create a log of `LogMarkerToken` instances.
* Chetan's proposal is to create `val`s in `CommonLoadBalancer` using `LOADBALANCER_COMPLETION_ACK` and then use these `LogMarkerToken` instances to emit  metrics. This makes a lot of sense!
* At the same time, I will keep `CompletionAckType` and sub-types as well as `LOADBALANCER_COMPLETION_ACK` in object `LoggingMarkers` to define all possible metrics in a single place. With that approach, you do not need multiple steps to look up metrics.
* I will also streamline `CompletionAckType` a little bit to save boilerplate code.